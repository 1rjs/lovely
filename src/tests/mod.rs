use std::fs;

use insta::glob;

use crate::{checker::Checker, ir::IRGenerator, lexer::Lexer, parser::Parser};

#[test]
fn compiler_tests() {
    glob!("source_files/*.lv", |path| {
        let input = fs::read_to_string(path).unwrap();

        // test the lexer
        let lexer = Lexer::new(&input);
        let tokens = lexer.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);

        // test the parser
        let mut parser = Parser::new(&input);
        let ast = parser.parse().unwrap();
        insta::assert_debug_snapshot!(ast);

        // test the checker
        let mut checker = Checker::new();
        let checked_program = checker.check(&ast).unwrap();
        insta::assert_debug_snapshot!(checked_program);

        // test the IR
        let ir_generator = IRGenerator::new(checker.types);
        let ir = ir_generator.program_ir(&checked_program);
        let ir_string = ir
            .iter()
            .map(|b| b.to_string())
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(ir_string);
    });
}
