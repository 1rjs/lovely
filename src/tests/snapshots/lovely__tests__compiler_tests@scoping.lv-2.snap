---
source: src/tests/mod.rs
expression: ast
input_file: src/tests/source_files/scoping.lv
---
Program(
    [
        Expression {
            span: Span {
                start: 0,
                end: 52,
            },
            kind: VariableDecl {
                name: "y",
                value: Expression {
                    span: Span {
                        start: 5,
                        end: 52,
                    },
                    kind: Block(
                        [
                            Expression {
                                span: Span {
                                    start: 9,
                                    end: 41,
                                },
                                kind: VariableDecl {
                                    name: "x",
                                    value: Expression {
                                        span: Span {
                                            start: 14,
                                            end: 41,
                                        },
                                        kind: Block(
                                            [
                                                Expression {
                                                    span: Span {
                                                        start: 20,
                                                        end: 26,
                                                    },
                                                    kind: VariableDecl {
                                                        name: "x",
                                                        value: Expression {
                                                            span: Span {
                                                                start: 25,
                                                                end: 26,
                                                            },
                                                            kind: IntLiteral(
                                                                7,
                                                            ),
                                                        },
                                                        mutable: false,
                                                        ty: None,
                                                    },
                                                },
                                                Expression {
                                                    span: Span {
                                                        start: 31,
                                                        end: 37,
                                                    },
                                                    kind: Infix {
                                                        left: Expression {
                                                            span: Span {
                                                                start: 31,
                                                                end: 32,
                                                            },
                                                            kind: Ident(
                                                                "x",
                                                            ),
                                                        },
                                                        operator: Minus,
                                                        right: Expression {
                                                            span: Span {
                                                                start: 35,
                                                                end: 37,
                                                            },
                                                            kind: IntLiteral(
                                                                23,
                                                            ),
                                                        },
                                                    },
                                                },
                                            ],
                                        ),
                                    },
                                    mutable: false,
                                    ty: None,
                                },
                            },
                            Expression {
                                span: Span {
                                    start: 44,
                                    end: 50,
                                },
                                kind: Infix {
                                    left: Expression {
                                        span: Span {
                                            start: 44,
                                            end: 45,
                                        },
                                        kind: Ident(
                                            "x",
                                        ),
                                    },
                                    operator: Plus,
                                    right: Expression {
                                        span: Span {
                                            start: 48,
                                            end: 50,
                                        },
                                        kind: IntLiteral(
                                            50,
                                        ),
                                    },
                                },
                            },
                        ],
                    ),
                },
                mutable: false,
                ty: None,
            },
        },
        Expression {
            span: Span {
                start: 54,
                end: 59,
            },
            kind: Infix {
                left: Expression {
                    span: Span {
                        start: 54,
                        end: 55,
                    },
                    kind: Ident(
                        "y",
                    ),
                },
                operator: Minus,
                right: Expression {
                    span: Span {
                        start: 58,
                        end: 59,
                    },
                    kind: IntLiteral(
                        2,
                    ),
                },
            },
        },
    ],
)
